subsys/shell: Introduce the Sensor Shell Module

diegosueiro opened the pull request:
Adds a Shell Module for retrieving Sensor data. The following commands
were added:

sensor - Device commands
Options:
  -h, --help  :Show command help.
Subcommands:
  get            :<device_name> [chanel_idx]
  list           :List configured sensors
  list_channels  :<device_name>

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

nashif commented on drivers/sensor/sensor_shell.c:
```diff
@@ -0,0 +1,228 @@
+
+	if (dev != NULL) {
+		k_thread_access_grant(k_current_get(), dev, NULL);
+		if (Z_SYSCALL_DRIVER_SENSOR(dev, sample_fetch)) {
```
not a public API

andrewboie replied:
I think I understand the intent here, but this isn't the right way to go about it.

> Even having the CONFIG_USERSAPCE=y set when calling sensor_sample_fetch it seems that it is calling the _impl_sensor_sample_fetch directly instead of Z_SYSCALL_HANDLER(sensor_sample_fetch, dev) defined in the drivers/sensor/sensor_handlers.c file

Yes, that's correct, handler checks are only performed when called from user mode, not supervisor mode.

> Do you have another suggestion to validade if a device is a sensor

The solution would be to have the shell subsystem shell() thread run in user mode.
Right now, everything is running in supervisor mode which assumes that all inputs are sane, there's  at best incomplete checking done. This is really a problem with the shell subsystem. There are all kinds of ways you can hose the system with bad shell commands, not just incorrect driver names. As a debugging/demo/test subsystem, historically people have been fine with the shell behaving in this way.

I'd simply drop the validation you are trying to do here; in order to properly have a shell which is robust against bad input, the whole thing needs to be in user mode. That way, if you start making API calls with bad input, the existing handlers will catch it.

diegosueiro replied:
@andrewboie ,

Thanks for the explanation. I removed the checks and the dependency on USERSPACE.
But I really think that we should have the shell running in a more secure environment with extra API call checks.

If you have time, please test this shell module by calling the following command: `sensor get UART_0` and `sensor get I2C_0`. I don't think this is is an acceptable behavior.

I tried to find in the shell subsys where it is dispatching the commands as threads but I couldn't find. Maybe we make use of the `k_thread_user_mode_enter`. But I'm not sure how complex can it be.

andrewboie replied:
> But I really think that we should have the shell running in a more secure environment with extra API call checks.

I understand where you are coming from but I think the only way to really assert security of the shell is to run it in user mode. Imagine if the Linux kernel console had an interactive command prompt that ran in kernel context...

>  I don't think this is is an acceptable behavior.

It depends. If all we are using shell for is debugging or demonstrations, then we don't really care about the security. If this is an end user facing feature, then it would be insane to expose it in its current form, and not just for being able to pass in bad driver names.

From my point of view, guarding against bad driver names for this specific command without actually running the entire shell in a reduced CPU privilege level is like locking the door and leaving all the windows open. There are so many shell commands implemented, any one of them could be an attack surface.

> I tried to find in the shell subsys where it is dispatching the commands as threads

Given that the shell subsys is processing arbitrary user input, I think it would only do to have the main shell thread started in shell.c start with K_USER options. It won't be a simple as that, there will be a lot more work to do since the shell subsystem was written before user mode was a part of Zephyr.

diegosueiro replied:
@andrewboie ,

I was thinking about creating an app that interacts with the user by printing and receiving data from UART. Having the shell subsys seems to be perfect for that. Lots of devices in the industrial field only have a serial port to interact with the user.

Do you think we should file an issue to get the shell subsys running in user mode and see what the others developers say?

tyrantlizard77 reviewed (dismissed):
-1 for reasons stated inline

andrewboie commented on samples/subsys/shell/sensor_module/src/main.c:
```diff
@@ -0,0 +1,14 @@
+
+void main(void)
+{
+
```
is this sample supposed to be doing something? or is main() blank because we just interact with the shell prompt?

diegosueiro replied:
The main() blank because we just interact with the shell prompt. Should I get rid of this?

diegosueiro commented:
@nashif ,

Any more comments?

nashif commented on drivers/sensor/sensor_shell.c:
```diff
@@ -0,0 +1,223 @@
+
+	shell_fprintf(shell, SHELL_NORMAL, "devices:\r\n");
+	for (dev = __device_init_start; dev != __device_init_end; dev++) {
+		k_thread_access_grant(k_current_get(), dev, NULL);
```
why do we still need this?

nashif commented on samples/subsys/shell/sensor_module/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 3.8.2)
```
sample needs to come as a separate commit

nashif reviewed (changes requested):
we need both a yaml file with a test and documentation file (README.rst) for the sample.

diegosueiro committed:
samples/subsys/shell: Introduce Sensor Shell module

Adds the Sensor Shell module sample app.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

diegosueiro commented:
> Found the following issues, please fix and resubmit:
> 
> ## Documentation Issues
> ```
> /var/lib/shippable/build/IN/main_repo/gitRepo/doc/_build/rst/doc/samples/subsys/shell/sensor_module/README.rst: WARNING: document isn't included in any toctree
> ```

@dbkinder,

I think I solved the documentation issue, but the test didn't run against my last commit.

erwango commented:
@avisconti , you might be interested

avisconti commented:
@erwango 
Indeed! It seems a good sensor abstraction layer. I will take a deeper look later on.

dbkinder commented on samples/subsys/shell/sensor_module/README.rst:
```diff
@@ -0,0 +1,94 @@
+         channel idx=3 accel_xyz =   1.991975
+         channel idx=3 accel_xyz =  -9.576807
+
+NOTE: a valid sensor device name should be passed otherwise you will get an
```
This would be better using a note directive (here and above on line 71):
```
.. note:: A valid sensor device name must be passed, otherwise you will get
   undefined behavior such as a hardware exception. The shell
   subsystem runs in supervisor mode where API callbacks are not checked before
   being called.
```

dbkinder commented on samples/subsys/shell/sensor_module/README.rst:
```diff
@@ -0,0 +1,94 @@
+.. _sensor_shell_sample:
```
You're getting a doc build error (this file is not included in any toctree).  Update your PR to fix this by adding a file ``samples/subsys/shell/shell.rst`` that contains:
```
.. _shell-samples:

Shell Samples
#############

.. toctree::
   :maxdepth: 1
   :glob:

   **/*
```
This .rst file is used to organize and pull in sample documentation for this subsystem (just like the other directories in this subsys section).

dbkinder reviewed (approved):
+1 for docs

diegosueiro commented:
@nashif,

Anything more is needed?

jakub-uC reviewed (changes requested):
I have proposed some suggestions regarding macros and commands registration.
However what you have to consider is that `shell_cmd_precheck` is no longer available as shell API.
As a result your code as is will not compile.

diegosueiro commented:
@jarz-nordic,

Thanks for the review. All comments addressed.

jarz-nordic commented on drivers/sensor/sensor_shell.c:
```diff
@@ -0,0 +1,195 @@
+		}
+
+	} else {
+		shell_error(shell, "Could not bind the %s sensor", argv[1]);
```
shouldn't you return an error code here?

jarz-nordic commented on drivers/sensor/sensor_shell.c:
```diff
@@ -0,0 +1,195 @@
+				shell_error(shell,
+					"Could not read channel idx %d=%s",
+					idx, sensor_channel_name[idx]);
+				return 1;
```
`1` is usually returned by shell command if help has been printed. Maybe you can use `-ENXIO` ? It is just a proposal I am not expert in POSIX error codes ;)

jarz-nordic commented on drivers/sensor/sensor_shell.c:
```diff
@@ -0,0 +1,195 @@
+		}
+
+	} else {
+		shell_error(shell, "Could not bind the %s sensor", argv[1]);
```
I think that here some error code should be returned.

diegosueiro committed:
subsys/shell: Introduce the Sensor Shell Module

Adds a Shell Module for retrieving Sensor data. The following commands
were added:

sensor - Device commands
Options:
  -h, --help  :Show command help.
Subcommands:
  get            :<device_name> [chanel_idx]
  list           :List configured sensors
  list_channels  :<device_name>

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented:
@jarz-nordic,

New comments addressed.

jakub-uC reviewed (approved):
From shell perspective LGTM

nashif commented:
sorry for the 1 year delay :)
